src/main/scala/chisel3/util/experimental/LoadMemoryTransform.scala

```diff
@@ -0,0 +1,231 @@
+// See LICENSE for license details.
+
+package chisel3.util.experimental
+
+import chisel3._
+import chisel3.experimental.annotate
+import chisel3.InstanceId
+import chisel3.experimental.{ChiselAnnotation, RunFirrtlTransform}
+import firrtl.annotations.{MemoryLoadFileType, _}
+import firrtl.ir.{Module => _, _}
+import firrtl.transforms.BlackBoxInlineAnno
+import firrtl.Mappers._
+import firrtl.{AnnotationSeq, CircuitForm, CircuitState, EmitCircuitAnnotation, LowForm, Transform, VerilogEmitter}
+
+import scala.collection.mutable
+
+/** Use this annotation to load a memory from a text file by using verilator and verilog's $readmemh or $readmemb
+  * The treadle backend can also recognize this annotation and load memory at run-time.
+  *
+  * This annotation triggers the [[LoadMemoryTransform]] which will take add the verilog directive to
+  * the relevant module by using the creating separate modules that are bound to the modules containing
+  * the memories to be loaded.
+  *
+  * ==Example module==
+  *
+  * Consider a simple Module containing a memory
+  * {{{
+  * import chisel3._
+  * class UsesMem(memoryDepth: Int, memoryType: Data) extends Module {
+  *   val io = IO(new Bundle {
+  *     val address = Input(UInt(memoryType.getWidth.W))
+  *     val value   = Output(memoryType)
+  *   })
+  *   val memory = Mem(memoryDepth, memoryType)
+  *   io.value := memory(io.address)
+  * }
+  * }}}
+  *
+  * ==Above module with annotation==
+  *
+  * To load this memory from a file /workspace/workdir/mem1.hex.txt
+  * Just add an import and annotate the memory
+  * {{{
+  * import chisel3._
+  * import chisel3.util.experimental.loadMemoryFromFile   // <<-- new import here
+  * class UsesMem(memoryDepth: Int, memoryType: Data) extends Module {
+  *   val io = IO(new Bundle {
+  *     val address = Input(UInt(memoryType.getWidth.W))
+  *     val value   = Output(memoryType)
+  *   })
+  *   val memory = Mem(memoryDepth, memoryType)
+  *   io.value := memory(io.address)
+  *   loadMemoryFromFile(memory, "/workspace/workdir/mem1.hex.txt")  // <<-- Note the annotation here
+  * }
+  * }}}
+  *
+  * ==Example file format==
+  * A memory file should consist of ascii text in either hex or binary format
+  * Example (a file containing the decimal values 0, 7, 14, 21):
+  * {{{
+  *   0
+  *   7
+  *   d
+  *  15
+  * }}}
+  * Binary file is similarly constructed.
+  *
+  * ==More info==
+  * See the LoadMemoryFromFileSpec.scala in the test suite for more examples
+  * @see <a href="https://github.com/freechipsproject/chisel3/wiki/Chisel-Memories">Load Memories in the Chisel wiki</a>
+  *
+  * @param target        memory to load
+  * @param fileName      name of input file
+  * @param hexOrBinary   use $readmemh or $readmemb, i.e. hex or binary text input, default is hex
+  */
+case class ChiselLoadMemoryAnnotation(
+  target:      InstanceId,
+  fileName:    String,
+  hexOrBinary: MemoryLoadFileType.FileType = MemoryLoadFileType.Hex
+)
```

Comment:
I think it got lost in Github's aggressive "hide outdated conversations 'feature'", but here is a diff that makes this more type safe and removes the need for `InstanceId`:
```diff
diff --git a/src/main/scala/chisel3/util/experimental/LoadMemoryTransform.scala b/src/main/scala/chisel3/util/experimental/LoadMemoryTransform.scala
index a81129d4..49a00637 100644
--- a/src/main/scala/chisel3/util/experimental/LoadMemoryTransform.scala
+++ b/src/main/scala/chisel3/util/experimental/LoadMemoryTransform.scala
@@ -4,7 +4,6 @@ package chisel3.util.experimental
 
 import chisel3._
 import chisel3.experimental.annotate
-import chisel3.InstanceId
 import chisel3.experimental.{ChiselAnnotation, RunFirrtlTransform}
 import firrtl.annotations.{MemoryLoadFileType, _}
 import firrtl.ir.{Module => _, _}
@@ -73,8 +72,8 @@ import scala.collection.mutable
   * @param fileName      name of input file
   * @param hexOrBinary   use $readmemh or $readmemb, i.e. hex or binary text input, default is hex
   */
-case class ChiselLoadMemoryAnnotation(
-  target:      InstanceId,
+case class ChiselLoadMemoryAnnotation[T <: Data](
+  target:      MemBase[T],
   fileName:    String,
   hexOrBinary: MemoryLoadFileType.FileType = MemoryLoadFileType.Hex
 )
@@ -95,8 +94,8 @@ case class ChiselLoadMemoryAnnotation(
 
 
 object loadMemoryFromFile {
-  def apply(
-    memory: MemBase[_],
+  def apply[T <: Data](
+    memory: MemBase[T],
     fileName: String,
     hexOrBinary: MemoryLoadFileType.FileType = MemoryLoadFileType.Hex
   ): Unit = {
```
Note this diff is slightly different than the previous one, I updated it with the recent changes so that it would apply cleanly

Reply:
done